### PR TITLE
Add question-mark-used lint

### LINKED ISSUE #10340: Deny Question Mark / Try Operator Lint
### What it does

I have a project where the try operator / question mark operator `?` is banned for performance reasons. I know that longer term ideally that was not necessary, but right now it is. Because of this I would like to avoid accidentally reintroducing it.

There are also other cases where having the ability to deny the question mark operator locally is useful. For instance there are cases where local information has to be attached to errors returned which is why you sometimes see constructs like this:

```rust
macro_rules! bail {
    ($err:expr) => {{
        let mut err = $err;
        process_err(&mut err, state);
        return Err(err);
    }};
}

macro_rules! ctx_ok {
    ($expr:expr) => {
        match $expr {
            Ok(rv) => rv,
            Err(err) => bail!(err),
        }
    };
}
```

In that case it would nice to be able to use `#[deny(clippy::use_of_try_operator)]` on a block to disallow it.

### Lint Name

use_of_try_operator

### Category

restriction

### Advantage

This is a lint that should never be turned on by default, it's only for specific code styles or local restrictions.

### Drawbacks

It might be turned on accidentally by developers not sure of it's actual intention.

### Example

```rust
let x = foo?;
```

Could be written as:

```rust
let x = ok!(foo);
```

### PR DESCRIPTION
**Pull request by mitsuhiko**
This lint complains when the question mark operator (try operator) is used.  This is a restriction lint that can be useful on local scopes where a custom error handling macro is supposed to be used to augment the error based on local scope data before returning.

Fixes #10340

---

changelog: New lint [`question_mark_used`]
[#10342](https://github.com/rust-lang/rust-clippy/pull/10342)

### CODE REVIEW

File: clippy_lints/src/question_mark_used.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    /// let result = expr?;
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```ignore
+    /// utility_macro!(expr);
+    /// ```
+    #[clippy::version = "pre 1.29.0"]
+    pub QUESTION_MARK_USED,
```
Comment (Alexendoo) [inline clippy_lints/src/question_mark_used.rs:27]: I agree with @Manishearth's [comment here](https://github.com/rust-lang/rust-clippy/issues/10340#issuecomment-1428173251), there are two other lints that refer to it as question mark currently but we could rename those as a separate issue

Comment (Manishearth) [reply clippy_lints/src/question_mark_used.rs:27]: No I don't have a strong opinion either way on question_mark vs try, I was only proposing a change of naming from `use_of_try_operator`. I think question_mark_used is fine

Comment (Manishearth) [reply clippy_lints/src/question_mark_used.rs:27]: The Rust Reference also just calls it "question mark", I think try is at this point a legacy name.

Comment (Alexendoo) [reply clippy_lints/src/question_mark_used.rs:27]: Ahh I see thanks. Looks like the book calls it question mark as well so the current name is good

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
Thank you!

@bors r+

**Pull request merged**